### PR TITLE
Making tests pass on GitLab

### PR DESCRIPTION
**Pull request by Yoric**
Took a few days of investigation, but the gist of it seems to be:

1. We need to put temporary files in `/builds/$CI_PROJECT_PATH/tmp` instead of `/tmp`.
2. We need to connect to the Docker registry, rather than trusting the daemon's default behavior.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -341,6 +357,23 @@ impl Config {
...
 }
 
+/// Configurable directories for this test.
+#[derive(Debug, TypedBuilder, Deserialize)]
+pub struct Directories {
+    /// The root of the test.
+    ///
+    /// All temporary files and logs are created as subdirectories of this directory.
+    ///
+    /// If unspecified, `$TMPDIR/mx-tester` (or the platform equivalent).
```
Comment (Gnuxie) [inline src/lib.rs:367]: is $TMPDIR a thing?

File: src/lib.rs [hunk 2]
```diff
@@ -198,6 +199,20 @@ pub struct Config {
     #[builder(default)]
     /// The version of Synapse to use
     pub synapse: SynapseVersion,
+
+    #[serde(default)]
+    #[builder(default)]
+    /// Information for logging to a registry.
+    ///
+    /// May be overridden from the command-line.
```
Comment (Gnuxie) [inline src/lib.rs:207]: Does the readme say how to do override from the command-line? I assume that you almost always want to do that when working with CI?

File: src/lib.rs [hunk 3]
```diff
@@ -868,13 +900,33 @@ pub async fn up(docker: &Docker, config: &Config) -> Result<(), Error> {
             docker.is_network_up(&network_name).await?,
             "The network should now be up"
         );
+    } else {
+        debug!("Network {} already exists", network_name);
     }
 
+    match config.up {
```
Comment (Gnuxie) [inline src/lib.rs:907]: What order has been changed here? Are we just making sure the docker network has been created before we run the up scripts?

**Review by Gnuxie — APPROVED** [review]
Have a few questions but otherwise it is nice :smiley_cat:

File: src/lib.rs (hunk 1 above)
Comment (Yoric) [reply src/lib.rs:367]: Probably not :)

File: src/lib.rs (hunk 2 above)
Comment (Yoric) [reply src/lib.rs:215]: Fixed that!

File: src/lib.rs (hunk 3 above)
Comment (Yoric) [reply src/lib.rs:915]: That's the idea, yes.

**Pull request merged**